aggregator/follower_flush_mgr.go

```diff
@@ -0,0 +1,279 @@
+		flushTimesKeyFmt:      opts.FlushTimesKeyFmt(),
+		flushTimesStore:       opts.FlushTimesStore(),
+		maxNoFlushDuration:    opts.MaxNoFlushDuration(),
+		forcedFlushWindowSize: opts.ForcedFlushWindowSize(),
+		logger:                instrumentOpts.Logger(),
+		scope:                 scope,
+		doneCh:                doneCh,
+		lastFlushed:           nowFn(),
+		sleepFn:               time.Sleep,
+		metrics:               newFollowerFlushManagerMetrics(scope),
+	}
+}
+
+func (mgr *followerFlushManager) Open(shardSetID string) {
+	mgr.Lock()
+	defer mgr.Unlock()
+
+	mgr.flushTimesKey = fmt.Sprintf(mgr.flushTimesKeyFmt, shardSetID)
+	go mgr.watchFlushTimes()
+}
+
+// NB(xichen): no actions needed for initializing the follower flush manager.
+func (mgr *followerFlushManager) Init(*sync.RWMutex, []*flushBucket) {}
+
+func (mgr *followerFlushManager) Flush(
+	bucketLock *sync.RWMutex,
+	buckets []*flushBucket,
+) (bool, time.Duration) {
+	var (
+		shouldFlush        = false
+		flushersByInterval []flushersGroup
+	)
+	// NB(xichen): a flush is triggered in the following scenarios:
+	// * The flush times persisted in kv have been updated since last flush.
+	// * Sufficient time (a.k.a. maxNoFlushDuration) has elapsed since last flush.
+	bucketLock.RLock()
+	mgr.Lock()
+	if mgr.flushTimesUpdated {
+		mgr.metrics.kvUpdateFlush.Inc(1)
+		shouldFlush = true
+		mgr.flushTimesUpdated = false
+		flushersByInterval = mgr.flushersFromKVUpdate(buckets)
+	} else {
+		now := mgr.nowFn()
+		durationSinceLastFlush := now.Sub(mgr.lastFlushed)
+		if durationSinceLastFlush > mgr.maxNoFlushDuration {
+			mgr.metrics.forcedFlush.Inc(1)
+			shouldFlush = true
+			flushBeforeNanos := now.Add(-mgr.maxNoFlushDuration).Add(mgr.forcedFlushWindowSize).UnixNano()
+			flushersByInterval = mgr.flushersFromForcedFlush(buckets, flushBeforeNanos)
+		}
+	}
+	mgr.Unlock()
+	bucketLock.RUnlock()
+
+	if !shouldFlush {
+		return false, mgr.checkEvery
+	}
+
+	var wgWorkers sync.WaitGroup
+	for _, group := range flushersByInterval {
+		start := mgr.nowFn()
+		for _, flusherWithTime := range group.flushers {
+			flusherWithTime := flusherWithTime
+			wgWorkers.Add(1)
+			mgr.workers.Go(func() {
+				flusherWithTime.flusher.FlushBefore(flusherWithTime.flushBeforeNanos)
+				wgWorkers.Done()
+			})
+		}
+		wgWorkers.Wait()
+		group.duration.Record(mgr.nowFn().Sub(start))
+	}
+	return true, 0
+}
+
+// NB(xichen): the follower flush manager flushes data based on the flush times
+// stored in kv and does not need to take extra actions when a new bucket is added.
+func (mgr *followerFlushManager) OnBucketAdded(int, *flushBucket) {}
+
+func (mgr *followerFlushManager) flushersFromKVUpdate(buckets []*flushBucket) []flushersGroup {
+	flushersByInterval := make([]flushersGroup, len(buckets))
+	for i, bucket := range buckets {
+		flushersByInterval[i].interval = bucket.interval
+		flushersByInterval[i].duration = bucket.duration
+		flushersByInterval[i].flushers = make([]flusherWithTime, 0, 16)
```

Comment:
Sure sounds good. Picked `16` as a default small size for upfront allocation because each node in prod owns roughly 15 shards.